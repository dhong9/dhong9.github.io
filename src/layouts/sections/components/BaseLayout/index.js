/*
=========================================================
* Material Kit 2 React - v2.1.0
=========================================================

* Product Page: https://www.creative-tim.com/product/material-kit-react
* Copyright 2023 Creative Tim (https://www.creative-tim.com)

Coded by www.creative-tim.com

 =========================================================

* The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
*/

import { useContext, useState, useEffect } from "react";

// prop-types is a library for typechecking of props
import PropTypes from "prop-types";

// @mui material components
import Container from "@mui/material/Container";
import Grid from "@mui/material/Grid";

// Material Kit 2 React components
import MKAvatar from "components/MKAvatar";
import MKBox from "components/MKBox";
import MKTypography from "components/MKTypography";

// Material Kit 2 React examples
import DefaultNavbar from "examples/Navbars/DefaultNavbar";
import CenteredFooter from "examples/Footers/CenteredFooter";
import Breadcrumbs from "examples/Breadcrumbs";

// DH React components
import DHSnackbar from "components/DHSnackbar";

// Authentication
import SignIn from "layouts/pages/authentication/sign-in";
import SignUp from "layouts/pages/authentication/sign-up";
import SignOut from "layouts/pages/authentication/sign-out";
import Profile from "layouts/pages/authentication/profile";
import AuthContext from "context/AuthContext";

// Routes
import routes from "routes";

// Images
import defaultProfileImage from "assets/images/default_profile.jpg";

// Services
import { getUserProfile } from "services/accountsService";

function BaseLayout({ breadcrumb, title, children }) {
  const { user, profile } = useContext(AuthContext);

  // Sign out properties
  const [snackbarOpen, setSnackbarOpen] = useState(false);
  const [signoutSeverity, setSignoutSeverity] = useState("info");
  const [signoutMessage, setSignoutMessage] = useState("");
  const [profileImage, setProfileImage] = useState(defaultProfileImage);

  useEffect(() => {
    // If there is a user, then get their profile info
    if (user) {
      getUserProfile(user.user_id, ({ data: { image } }) => {
        setProfileImage(image);
      });
    }
  });

  const signoutSuccess = () => {
    setSignoutSeverity("success");
    setSignoutMessage("Successfully signed out!");
    setSnackbarOpen(true);
  };

  const handleSnackbarClose = (event, reason) => {
    if (reason !== "clickaway") {
      setSnackbarOpen(false);
    }
  };

  const signedOutOptions = [
    {
      name: "Sign In",
      route: "/pages/authentication/sign-in",
      component: <SignIn />,
    },
    {
      name: "Sign Up",
      route: "/pages/authentication/sign-up",
      component: <SignUp />,
    },
  ];
  const signedInOptions = [
    {
      name: "Profile",
      route: "/pages/authentication/profile",
      component: <Profile />,
    },
    {
      name: "Sign Out",
      route: "/pages/authentication/sign-out",
      component: <SignOut onload={signoutSuccess} />,
    },
  ];

  const accountObj = {
    // If the user is logged in, use their account's username
    name: user ? user.username || profile.name : "Guest",

    // If the user has a profile picture from Google, use that image
    // Otherwise, use the picture tied to the site's account
    icon: (
      <MKAvatar
        src={user && profile.picture ? profile.picture : profileImage}
        alt="Profile picture"
        size="xs"
      />
    ),
    collapse: user ? signedInOptions : signedOutOptions,
  };

  return (
    <>
      <DHSnackbar
        open={snackbarOpen}
        onClose={handleSnackbarClose}
        severity={signoutSeverity}
        message={signoutMessage}
      />
      <MKBox display="flex" flexDirection="column" bgColor="white" minHeight="100vh">
        <MKBox bgColor="white" shadow="sm" py={0.25}>
          <DefaultNavbar routes={[...routes, accountObj]} transparent relative />
        </MKBox>
        <Container sx={{ mt: 6 }}>
          <Grid container item xs={12} flexDirection="column" justifyContent="center" mx="auto">
            <MKBox width={{ xs: "100%", md: "50%", lg: "25%" }} mb={3}>
              <Breadcrumbs routes={breadcrumb} />
            </MKBox>
            <MKTypography variant="h3" mb={1}>
              {title}
            </MKTypography>
            {children}
          </Grid>
        </Container>
        <MKBox mt="auto">
          <CenteredFooter />
        </MKBox>
      </MKBox>
    </>
  );
}

// Typechecking props for the BaseLayout
BaseLayout.propTypes = {
  breadcrumb: PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.object])).isRequired,
  title: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired,
};

export default BaseLayout;
