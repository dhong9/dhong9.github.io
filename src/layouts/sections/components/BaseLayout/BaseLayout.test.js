import renderer from "react-test-renderer";
import BaseLayout from "layouts/sections/components/BaseLayout";
import React from "react";

let realUseContext;
let useContextMock;
// Setup mock
beforeEach(() => {
    realUseContext = React.useContext;
    useContextMock = React.useContext = jest.fn();
});
// Cleanup mock
afterEach(() => {
    React.useContext = realUseContext;
});

// Define Mocks
jest.mock("components/MKBox", () => {
    const { forwardRef } = jest.requireActual("react");
    return {
        __esModule: true,
        default: forwardRef(() => <div>Mock Box</div>)
    };
});
jest.mock("components/MKTypography", () => {
    const { forwardRef } = jest.requireActual("react");
    return {
        __esModule: true,
        default: forwardRef(() => <p>Mock Typography</p>)
    };
});
jest.mock("examples/Breadcrumbs", () => {
    const { forwardRef } = jest.requireActual("react");
    return {
        __esModule: true,
        default: forwardRef(() => <div>Mock Breadcrumbs</div>)
    };
});
jest.mock("react-monaco-editor", () => {
    const { forwardRef } = jest.requireActual("react");
    return {
        __esModule: true,
        default: forwardRef(() => <div>Mock Editor</div>)
    };
});
jest.mock("context/AuthContext", () => {
    const { createContext } = jest.requireActual('react'); 
    return {
        __esModule: true,
        default: createContext()
    }
});

describe("BaseLayout", () => {
    it("renders", () => {
        useContextMock.mockReturnValue("Test Value");
        const component = renderer.create(
            <BaseLayout
                title="Test Layout"
                breadcrumb={[]}
            >
                    Base
            </BaseLayout>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});