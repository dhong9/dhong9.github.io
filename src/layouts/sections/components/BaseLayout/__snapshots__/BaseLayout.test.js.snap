// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseLayout BaseLayout 1`] = `
<div>
  Mock Box
</div>
`;
